File: app/src/androidTestScreenshots/java/openfoodfacts/github/scrachx/openfood/test/ScreenshotActivityTestRule.java
package openfoodfacts.github.scrachx.openfood.test;

import android.app.Activity;
import android.content.Intent;
import android.util.Log;

import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.rule.ActivityTestRule;

import org.apache.commons.lang.StringUtils;
import org.junit.Assert;

import java.util.function.Consumer;

import openfoodfacts.github.scrachx.openfood.utils.LocaleHelper;
import openfoodfacts.github.scrachx.openfood.utils.PrefManager;
import openfoodfacts.github.scrachx.openfood.views.OFFApplication;

public class ScreenshotActivityTestRule<T extends Activity> extends ActivityTestRule<T> {
    public static final int MILLIS_TO_WAIT_TO_DISPLAY_ACTIVITY = 5000;
    private Intent activityIntent;
    private Consumer<ScreenshotActivityTestRule<T>> afterActivityLaunchedAction;
    private Consumer<ScreenshotActivityTestRule<T>> beforeActivityStartedAction;
    private boolean firstTimeLaunched = false;
    private String name;
    private ScreenshotParameter screenshotParameter;

    public ScreenshotActivityTestRule(Class<T> activityClass) {
        this(activityClass, activityClass.getSimpleName(), null);
    }

    public ScreenshotActivityTestRule(Class<T> activityClass, String name) {
        this(activityClass, name, null);
    }

    public ScreenshotActivityTestRule(Class<T> activityClass, Intent intent) {
        this(activityClass, activityClass.getSimpleName(), intent);
    }

    public ScreenshotActivityTestRule(Class<T> activityClass, String name, Intent intent) {
        super(activityClass, false, false);
        this.name = name;
        this.activityIntent = intent;
    }

    public void setBeforeActivityStartedAction(Consumer<ScreenshotActivityTestRule<T>> beforeActivityStartedAction) {
        this.beforeActivityStartedAction = beforeActivityStartedAction;
    }

    public ScreenshotParameter getScreenshotParameter() {
        return screenshotParameter;
    }

    public void setScreenshotParameter(ScreenshotParameter screenshotParameter) {
        this.screenshotParameter = screenshotParameter;
    }

    public void setAfterActivityLaunchedAction(Consumer<ScreenshotActivityTestRule<T>> afterActivityLaunchedAction) {
        this.afterActivityLaunchedAction = afterActivityLaunchedAction;
    }

    public void setFirstTimeLaunched(boolean firstTimeLaunched) {
        this.firstTimeLaunched = firstTimeLaunched;
    }

    @Override
    protected void beforeActivityLaunched() {
        try {
            runOnUiThread(() -> {
                new PrefManager(OFFApplication.getInstance()).setFirstTimeLaunch(firstTimeLaunched);
                LocaleHelper.setLocale(InstrumentationRegistry.getInstrumentation().getTargetContext(), screenshotParameter.getLocale());
            });
        } catch (Throwable throwable) {
            throwable.printStackTrace();
            Assert.fail(throwable.getMessage());
        }
        if (beforeActivityStartedAction != null) {
            beforeActivityStartedAction.accept(this);
        }
    }

    @Override
    protected void afterActivityLaunched() {
        try {
            if (afterActivityLaunchedAction != null) {
                afterActivityLaunchedAction.accept(this);
            }
            Thread.sleep(MILLIS_TO_WAIT_TO_DISPLAY_ACTIVITY);
            InstrumentationRegistry.getInstrumentation().waitForIdleSync();
            takeScreenshot();
            this.finishActivity();
        } catch (Throwable throwable) {
            Log.e(ScreenshotActivityTestRule.class.getSimpleName(), "run on ui", throwable);
        }
    }

    public void takeScreenshot() {
        takeScreenshot(StringUtils.EMPTY);
    }

    public void takeScreenshot(String suffix) {
        ScreenshotTaker taker = new ScreenshotTaker();
        taker.takeScreenshot(screenshotParameter, suffix, this);
    }

    @Override
    protected Intent getActivityIntent() {
        return activityIntent;
    }

    public void setActivityIntent(Intent activityIntent) {
        this.activityIntent = activityIntent;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}

File: app/src/androidTestScreenshots/java/openfoodfacts/github/scrachx/openfood/test/ScreenshotActivityTestRule.kt
package openfoodfacts.github.scrachx.openfood.test

import android.app.Activity
import android.util.Log
import androidx.test.core.app.ApplicationProvider
import androidx.preference.PreferenceManager
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.rule.ActivityTestRule
import openfoodfacts.github.scrachx.openfood.utils.LocaleManager
import openfoodfacts.github.scrachx.openfood.utils.PrefManager
import org.apache.commons.lang3.StringUtils
import org.junit.Assert

class ScreenshotActivityTestRule<T : Activity?>
@JvmOverloads constructor(
        activityClass: Class<T>,
        var name: String = activityClass.simpleName,
        val context: Context,
        private val localeManager: LocaleManager,
) : ActivityTestRule<T>(activityClass, false, false) {
    var afterActivityLaunchedAction: ((ScreenshotActivityTestRule<T>) -> Unit)? = null
    var beforeActivityStartedAction: ((ScreenshotActivityTestRule<T>) -> Unit)? = null
    var firstTimeLaunched = false
    var screenshotParameter: ScreenshotParameter? = null


    @Suppress("DEPRECATION")
    override fun beforeActivityLaunched() {
        try {
            runOnUiThread {
                PrefManager(ApplicationProvider.getApplicationContext()).isFirstTimeLaunch = firstTimeLaunched
                LocaleHelper.setContextLanguage(
                        InstrumentationRegistry.getInstrumentation().targetContext,
                        screenshotParameter!!.locale,
                )
            }
        } catch (throwable: Throwable) {
            throwable.printStackTrace()
            Assert.fail(throwable.message)
        }
        beforeActivityStartedAction?.invoke(this)
    }

    override fun afterActivityLaunched() {
        try {
            afterActivityLaunchedAction?.invoke(this)

            Thread.sleep(MILLIS_TO_WAIT_TO_DISPLAY_ACTIVITY.toLong())
            InstrumentationRegistry.getInstrumentation().waitForIdleSync()
            takeScreenshot()
            finishActivity()
        } catch (throwable: Throwable) {
            Log.e(ScreenshotActivityTestRule::class.simpleName, "run on ui", throwable)
        }
    }

    private fun takeScreenshot(suffix: String = StringUtils.EMPTY) {
        ScreenshotTaker.takeScreenshot(screenshotParameter!!, suffix, this)
    }

    companion object {
        const val MILLIS_TO_WAIT_TO_DISPLAY_ACTIVITY = 5000
    }

}
